server/webex_store.go

```
@@ -0,0 +1,270 @@
+	if err != nil {
+		return session, err
+	}
+
+	session.Token.AccessToken = accessToken
+
+	// Decrypt refresh token if it's present
+	if session.Token.RefreshToken != "" {
+
+		refreshToken, err := decrypt([]byte(config.EncryptionKey), session.Token.RefreshToken)
+		if err != nil {
+			return session, err
+		}
+
+		session.Token.RefreshToken = refreshToken
+	}
+
+	return session, nil
+}
+
+// getWebexUserInfo retrieves the webex user info from the KV store and
+// if not present will request from the webex API and subsequently save
+// it to the KV store if successful
+func (p *Plugin) getWebexUserInfo(userID string) (WebexUserInfo, error) {
+
+	webexUser, loadErr := p.loadWebexUser(userID)
+
+	switch {
+	case loadErr == ErrWebexUserNotFound:
+
+		session, err := p.loadWebexSession(userID)
+		if err != nil {
+			p.API.LogError("error looking up session", "error", err.Error())
+			return webexUser, err
+		}
+
+		webex, err := NewWebexClient(session.Token.AccessToken)
+		if err != nil {
+			p.API.LogError("Error creating new webex client", "error", err.Error())
+			return webexUser, err
+		}
+
+		// unlikely, but just in case to avoid angry panics
+		if webex == nil {
+			p.API.LogError("nil webex client")
+			return webexUser, errors.New("webex client is nil")
+		}
+
+		person, _, err := webex.People.GetMe() // don't need resp so supressing it
+		if err != nil {
+			p.API.LogError("Error calling people.GetMe()", "error", err.Error())
+			return webexUser, err
+		}
+
+		webexUser.FromWebexPerson(person)
+
+		err = p.storeWebexUser(userID, webexUser)
+		if err != nil {
+			p.API.LogError("Error saving webex user", "error", err.Error())
+			return webexUser, err
+		}
+
+	case (loadErr != nil):
+		return webexUser, loadErr
+	}
+
+	return webexUser, nil
+
+}
+
+// loadWebexUser loads the webex user info from the KV store if present
+func (p *Plugin) loadWebexUser(userID string) (WebexUserInfo, error) {
```

For this cases normally we return `(*WebexUserInfo, error)`, it is not super important, but is more clear in the code that you are returning `nil, err`, instead `webexUser, err` which is kinda confusing sometimes. Actually in certain situations I can want to return an error and a partial result, but in this cases we only use the `webexUser` if the error is `nil`.

---

Probably this is applicable to other functions.

---

Good point, I can update it to a reference.